src/run/mod.rs

```diff
@@ -59,60 +62,134 @@ pub fn run(args: &RunArgs) -> anyhow::Result<()> {
     Ok(())
 }
 
-/// Determine which package should be run.
+#[derive(Debug, Clone)]
+pub(crate) struct BinTarget {
+    /// The path to the directory in `target` which contains the binary.
+    pub(crate) artifact_directory: PathBuf,
+    /// The name of the binary (without any extensions).
+    pub(crate) bin_name: String,
+}
+
+/// Determine which binary target should be run.
 ///
-/// We first take a look at the `--bin` and `--package` args.
-/// If they are not defined, we try to determine the package automatically with the information
-/// provided by cargo metadata.
-/// We first look for the `default_run` definition and otherwise check if there is only a single
-/// binary package that could be run.
-fn select_run_package<'a>(
-    metadata: &'a Metadata,
-    args: &CargoRunArgs,
-) -> anyhow::Result<&'a Package> {
-    let package_name = if let Some(bin) = &args.target_args.bin {
-        bin.clone()
-    } else if let Some(package) = &args.package_args.package {
-        package.clone()
-    } else {
-        // Try to determine the run package automatically
-        let default_runs: Vec<_> = metadata
+/// The `--package` arg narrows down the search space to the given package,
+/// while the `--bin` and `--example` args determine the binary target within the selected packages.
+///
+/// If the search couldn't be narrowed down to a single binary,
+/// the `default_run` option is taken into account.
+///
+/// The path to the compiled binary is determined via the compilation target and profile.
+pub(crate) fn select_run_binary(
+    metadata: &Metadata,
+    package_name: &Option<String>,
+    bin_name: &Option<String>,
+    example_name: &Option<String>,
+    compile_target: &Option<String>,
```

Comment:
I think these can all be `Option<&str>` which performs better, and more importantly feels nicer.

It can also totally be left to a follow up.

Reply:
Very good point, changed this and used `.deref()` to adjust the call sites